File: packages/shared/src/vinValidation.ts
// Canonical VIN/HIN identifier validation (ISO 3779 check-digit algorithm).
// Previously duplicated byte-for-byte between packages/web and
// packages/functions — moved here as the single source. Kept as an
// isomorphic module rather than a callable: this needs to run at
// keystroke speed for instant form feedback, and a network round-trip
// here would be actively harmful, not just unnecessary latency.

const VIN_TRANSLITERATION: Record<string, number> = {
  A: 1,
  B: 2,
  C: 3,
  D: 4,
  E: 5,
  F: 6,
  G: 7,
  H: 8,
  J: 1,
  K: 2,
  L: 3,
  M: 4,
  N: 5,
  P: 7,
  R: 9,
  S: 2,
  T: 3,
  U: 4,
  V: 5,
  W: 6,
  X: 7,
  Y: 8,
  Z: 9,
};

const VIN_WEIGHTS = [8, 7, 6, 5, 4, 3, 2, 10, 0, 9, 8, 7, 6, 5, 4, 3, 2];

export type VehicleIdentifierType = 'empty' | 'vin' | 'hin' | 'serial';

function looksLikeVin(vinInput: string): boolean {
  const vin = (vinInput || '').trim().toUpperCase();
  return /^[A-HJ-NPR-Z0-9]{17}$/.test(vin);
}

export function hasValidHinFormat(hinInput: string): boolean {
  const hin = (hinInput || '').trim().toUpperCase();
  return /^[A-HJ-NPR-Z]{3}[A-HJ-NPR-Z0-9]{9}$/.test(hin);
}

export function detectVehicleIdentifierType(
  identifierInput: string,
  vehicleTypeInput?: string
): VehicleIdentifierType {
  const identifier = (identifierInput || '').trim().toUpperCase();
  const vehicleType = (vehicleTypeInput || '').trim().toLowerCase();

  if (!identifier) {
    return 'empty';
  }

  if (looksLikeVin(identifier)) {
    return 'vin';
  }

  if (hasValidHinFormat(identifier)) {
    return 'hin';
  }

  if (vehicleType.includes('boat') && identifier.length === 12) {
    return 'hin';
  }

  return 'serial';
}

export function hasValidVinChecksum(vinInput: string): boolean {
  const vin = (vinInput || '').trim().toUpperCase();

  if (!looksLikeVin(vin)) {
    return false;
  }

  let sum = 0;
  for (let i = 0; i < vin.length; i += 1) {
    const char = vin[i];
    const numericValue = Number.isFinite(Number(char))
      ? Number(char)
      : VIN_TRANSLITERATION[char];

    if (numericValue === undefined) {
      return false;
    }

    sum += numericValue * VIN_WEIGHTS[i];
  }

  const remainder = sum % 11;
  const expectedCheckDigit = remainder === 10 ? 'X' : String(remainder);
  return vin[8] === expectedCheckDigit;
}

export function getVinLookupValidationError(vinInput: string): string | null {
  const vin = (vinInput || '').trim().toUpperCase();

  if (!looksLikeVin(vin)) {
    return 'VIN lookup requires a valid 17-character VIN.';
  }

  if (!hasValidVinChecksum(vin)) {
    return 'VIN lookup requires a valid 17-character VIN with a correct check digit.';
  }

  return null;
}

File: packages/mobile/lib/utils/vin_validation.dart
// Mobile port of packages/shared/src/vinValidation.ts (ISO 3779 VIN
// check-digit algorithm). Previously mobile only checked VIN length
// (17 characters) with no checksum validation at all, unlike web/Functions
// — this closes that gap.

const Map<String, int> _vinTransliteration = {
  'A': 1,
  'B': 2,
  'C': 3,
  'D': 4,
  'E': 5,
  'F': 6,
  'G': 7,
  'H': 8,
  'J': 1,
  'K': 2,
  'L': 3,
  'M': 4,
  'N': 5,
  'P': 7,
  'R': 9,
  'S': 2,
  'T': 3,
  'U': 4,
  'V': 5,
  'W': 6,
  'X': 7,
  'Y': 8,
  'Z': 9,
};

const List<int> _vinWeights = [8, 7, 6, 5, 4, 3, 2, 10, 0, 9, 8, 7, 6, 5, 4, 3, 2];

final RegExp _vinPattern = RegExp(r'^[A-HJ-NPR-Z0-9]{17}$');
final RegExp _hinPattern = RegExp(r'^[A-HJ-NPR-Z]{3}[A-HJ-NPR-Z0-9]{9}$');

bool looksLikeVin(String value) {
  return _vinPattern.hasMatch(value.trim().toUpperCase());
}

bool hasValidHinFormat(String value) {
  return _hinPattern.hasMatch(value.trim().toUpperCase());
}

String detectVehicleIdentifierType(String identifierInput, [String? vehicleTypeInput]) {
  final identifier = identifierInput.trim().toUpperCase();
  final vehicleType = (vehicleTypeInput ?? '').trim().toLowerCase();

  if (identifier.isEmpty) {
    return 'empty';
  }

  if (looksLikeVin(identifier)) {
    return 'vin';
  }

  if (hasValidHinFormat(identifier)) {
    return 'hin';
  }

  if (vehicleType.contains('boat') && identifier.length == 12) {
    return 'hin';
  }

  return 'serial';
}

bool hasValidVinChecksum(String value) {
  final vin = value.trim().toUpperCase();

  if (!looksLikeVin(vin)) {
    return false;
  }

  var sum = 0;
  for (var i = 0; i < vin.length; i += 1) {
    final char = vin[i];
    final digit = int.tryParse(char);
    final numericValue = digit ?? _vinTransliteration[char];

    if (numericValue == null) {
      return false;
    }

    sum += numericValue * _vinWeights[i];
  }

  final remainder = sum % 11;
  final expectedCheckDigit = remainder == 10 ? 'X' : remainder.toString();
  return vin[8] == expectedCheckDigit;
}

String? getVinLookupValidationError(String value) {
  final vin = value.trim().toUpperCase();

  if (!looksLikeVin(vin)) {
    return 'VIN lookup requires a valid 17-character VIN.';
  }

  if (!hasValidVinChecksum(vin)) {
    return 'VIN lookup requires a valid 17-character VIN with a correct check digit.';
  }

  return null;
}
